### DISCUSSION TITLE
ChangeLogs

### DISCUSSION THREAD

**Discussion by smly** [smly/mjai.app#28] [Announcements]
This thread is for reporting minor changes.

## 2023-08-06 (Sun) 22:00 JST
* Implemented a button to move to the next tsumo event.
* Implemented permalink to move to a given kyoku and step number. E.g.: [mjai.app/games/2/250/2?k=10&s=119](https://mjai.app/games/2/250/2?k=10&s=119)

## 2023-08-06 (Sun) 14:00 JST
  * Updated ratings by running weekly batch.
  * Added the ability to browse games and standings for each batch in the web UI ([mjai.app/games](https://mjai.app/games)).
  * Next: Will focus on documentation and SDK development.

Many features are still under development, so ideas and bug reports are very welcome. Feel free to create a new thread.

Thanks,

**Comment by smly** [comment]
## 2023-08-13 (Sun) 01:30 JST: Refactoring & High-level API

* Merged https://github.com/smly/mjai.app/pull/30
* Added rule-based bot using high-level API https://github.com/smly/mjai.app/blob/main/examples/rulebase/bot.py
* Updated docker image `docker.io/smly/mjai-client:v3`.
* Updated submission server. Verified that [rulebase.zip](https://github.com/smly/mjai.app/blob/main/examples/rulebase.zip) passes validation in [mjai.app/submissions](https://mjai.app/submissions).
* Basic API implemented. Still not implemented kan actions, etc. Will be addressed in the following days.

I realize that debugging is difficult from the limited information available, so I may add a feature for debugging.

## 2023-08-11 (Fri) 14:40 JST: Refactoring & High-level API

* [WIP] Refactoring: https://github.com/smly/mjai.app/pull/30
    * Rename `mjaisimulator` module to `mjai` module
    * Add `Bot` class as high-level API to `mjai.bot` module to make it easier to develop bots

The `mjai` package is being considered to ship to PyPI.
Below is an example implementation using the new high-level API. The return value of `think()` method is a serialized event message.

```py
import sys
from mjai.bot import Bot

class RiichiBot(Bot):
    def think(self) -> str:
        if self.can_tsumo_agari:
            return self.action_tsumo_agari()
        elif self.can_ron_agari:
            return self.action_ron_agari()
        elif self.can_riichi:
            return self.action_riichi()
        elif self.can_discard:
            candidates = self.find_improving_tiles()
            for candidate in candidates:
                return self.action_discard(candidate["discard_tile"])
            return self.action_discard(
                self.last_self_tsumo or self.tehai_mjai[0]
            )
        else:
            # Response toward start_game, ryukyoku, etc
            return self.action_nothing()

if __name__ == "__main__":
    RiichiBot(player_id=int(sys.argv[1])).start()
```

**Comment by smly** [comment]
## 2023-08-14 (Mon) 05:00 - 21:00 JST: Bug fixes & High-level API

* Impl high-level API
  * https://github.com/smly/mjai.app/pull/40
  * https://github.com/smly/mjai.app/pull/41
  * https://github.com/smly/mjai.app/pull/39
  * https://github.com/smly/mjai.app/pull/34
* Example code with high-level API
  * https://github.com/smly/mjai.app/pull/35
* mjai.app simulator
  * https://github.com/smly/mjai.app/pull/33 and https://github.com/smly/mjai.app/pull/34/files#r1292848338

**Comment by smly** [comment]
## 2023-08-16 (Tue) 07:00 JST: Many updates on high-level API & web UI

* Update Docker with latest `mjai` package ([tag:v20230816](https://github.com/smly/mjai.app/releases/tag/v20230816)). 
  * Digest: `51e413b45903` (https://hub.docker.com/r/smly/mjai-client/tags)
* Update high-level API
  * https://github.com/smly/mjai.app/pull/46
  * https://github.com/smly/mjai.app/pull/47
  * https://github.com/smly/mjai.app/pull/49
* Many updates on Web UI
  * Changed to show score changes in game viewer
![image](https://github.com/smly/mjai.app/assets/19337/f38b768d-bc95-40f5-81f9-85cde3bcffd4)

**Comment by smly** [comment]
## 2023-09-04 (Mon): Enhancement in Game Viewer

### New features

* Added animations for discarding tiles in the game viewer
* Added button to filter games by ELO rating range

https://github.com/smly/mjai.app/assets/19337/718dc3d6-4ae0-4087-a27b-012195f37f98

### Weekly batch processing on Google Cloud

Ongoing. batch6 was originally run on GCE e2-standard-8. However, there was an issue that caused an extreme drop in ML bot ratings during the batch. The cause is under investigation. It is likely that the CPU was not powerful enough to handle the ML bot, resulting in many timeouts. I will continue to work on debugging this issue.

### Stability of ELO-rating

For better ELO-rating stability, I plan to improve the matching method to increase the number of matches between bots in the same ELO-rating range. I also plan to increase the number of matches in the higher ELO ranges.

**Reply by weiwangorg** [reply]
Would you open source the web UI? I feel that using the web UI to replay game logs would be more intuitive and easier to understand, and it would also conveniently allow human-model battles to better perceive the differences between bots and humans.

**Comment by smly** [comment]
## 2023-09-16 (Sat): Updated the matching algorithm

### Change of matching algorithm

Separated bots into two groups based on ELO score of weakml. Opponents will be selected from the same group. As the result, more matches were made for strong bots. On the other hand, strong bots have relatively long execution time. Each match can take 20 minutes or more. The batch process is expected to take more than 50 hours to complete, so the results may not be output by Monday.

### WIP: Results during the batch

The results of batch processing can now be checked during the process. Since it is still under development, there is no link from the top page yet.

Link: https://mjai.app/games/8/leaderboard

### Game view for smartphones

Added game view for smartphones.

![image](https://github.com/smly/mjai.app/assets/19337/07a236b1-5c38-434e-a0c4-7e824785913d)

**Comment by smly** [comment]
## 2024-04-08 (Mon): Debugging

* (2024-04-08 23:07) Bugfix: weekly batch process does not start correctly when docker container needs time to start. Changed the time to wait for docker container to start from 1 second to 5 seconds.
* (2024-04-09 03:00) It should be bankrupt (having less than 0 points) but the score is wrong. https://mjai.app/games/37/5/0?k=15&s=85
* (2024-04-10 04:40) Updated dependent packages and uploaded `mjai` package to [PyPI](https://pypi.org/project/mjai/).

**Comment by smly** [comment]
## 2024-04-19 (Fri): Moved batch processing from home to a cloud

In the future, the use of many machines on the cloud platform may reduce the computation time for our weekly batch processing.
